### PR TITLE
Support user desktop wallpapers #1326

### LINKED ISSUE #1326: Add support to display via a lightdm based greeter to show the selected user chosen wallpaper
Issue raised from here https://github.com/linuxmint/slick-greeter/issues/94#issuecomment-365616160

Applies to lightdm v1.25.2 and later

Recent changes in [lightdm](https://github.com/CanonicalLtd/lightdm/commit/6015bce25f241e7580c03594d846769f8236232f) now support the capability of a lightdm greeter (such as slick greeter) to display the wallpaper of the current highlighted user on the greeter.  Its a bit of bling that was previously limited to Ubuntu based distro's.    It didnt apply to encrypted home-drives I understand.

The changes to lightdm (according to the maintainer) utilise upstream AccountsService changes and thus enable all distros to benefit from this capability.

It is entirely up-to the distro itself whether to enable/disable the capability via the lightdm config file.

The suggestion here is to add the callout to the dbus service mentioned in the linked issue.  First off the blocks were linux mint with cinnamon.  Implementation was via their [settings-damon](https://github.com/linuxmint/cinnamon-settings-daemon/commit/63f501e60cbff2a8989e29dfb65018db8269e0a7) by calling gnome_bg_set_accountsservice_background whenever the gnome wallpaper settings is changed and also when first drawing the background (I think).

@ikeydoherty mentioned possibly adding such capability to budgie-daemon.

Speculation also on the budgie implementation: 

Maybe also to https://github.com/budgie-desktop/budgie-desktop/blob/master/src/wm/background.vala ? i.e. line 70 for the change event on settings and maybe the function at line 206 - both add the same callout to gnome_bg.set_accountsservice_background (I'm assuming that isnt a cinnamon specific function) ?

### PR DESCRIPTION
**Pull request by fossfreedom**
Resolves #1326 

Testing Performed with lightdm/slick-greeter

1. Choosing an XML based wallpaper via GCC - no change on slick-greeter/no warnings in journalctl
2. Choosing a valid wallpaper via GCC - change on slick-greeter/no warnings in journalctl
3. Switching off User Backgrounds via lightdm-settings and choosing a valid wallpaper via GCC - no change on slick-greeter/no warnings in journalctl
4. Switching on User Backgrounds via lightdm-settings and settings a blank wallpaper (empty string) via dconf - slick-greeter changed to default background, no warnings in journalctl

Switched to GDM3

1. Choose a valid wallpaper via GCC - no change to GDM background/no warnings in journalctl

### CODE REVIEW

**Comment by Jacalz** [comment]
Would really like seeing this merged, a great feature to have in Budgie :+1:

File: src/wm/background.vala [hunk 1]
```diff
@@ -140,6 +140,71 @@ public class BudgieBackground : Meta.BackgroundGroup
...
+        catch (IOError e)
+        {
+            warning("Failed to get system bus: %s", e.message);
+            return;
+        }
+
+        try
+        {
+            variant = bus.call_sync(
+                "org.freedesktop.Accounts",
```
Comment (JoshStrobl) [inline src/wm/background.vala]: Can just create a schema const to reference like `ACCOUNTS_SCHEMA` and save the multiple identical strings.

File: src/wm/background.vala [hunk 2]
```diff
@@ -140,6 +140,71 @@ public class BudgieBackground : Meta.BackgroundGroup
...
+        }
+        catch (IOError e)
+        {
+            warning("Failed to get system bus: %s", e.message);
+            return;
+        }
+
+        try
+        {
+            variant = bus.call_sync(
```
Comment (JoshStrobl) [inline src/wm/background.vala]: Doesn't really need to be split up into a million lines.

File: src/wm/background.vala [hunk 3]
```diff
@@ -140,6 +140,71 @@ public class BudgieBackground : Meta.BackgroundGroup
...
+            warning("Could not contact accounts service to look up '%s': %s", Environment.get_user_name(), e.message);
+            return;
+        }
+
+        string object_path = variant.get_child_value(0).get_string();
+
+        try
+        {
+            bus.call_sync(
+                "org.freedesktop.Accounts",
```
Comment (JoshStrobl) [inline src/wm/background.vala]: Create + use an ACCOUNT_SCHEMA const to reduce redundancy.

File: src/wm/background.vala [hunk 4]
```diff
@@ -140,6 +140,71 @@ public class BudgieBackground : Meta.BackgroundGroup
...
+        {
+            warning("Could not contact accounts service to look up '%s': %s", Environment.get_user_name(), e.message);
+            return;
+        }
+
+        string object_path = variant.get_child_value(0).get_string();
+
+        try
+        {
+            bus.call_sync(
```
Comment (JoshStrobl) [inline src/wm/background.vala]: Doesn't really need to be split up into a million lines.

File: src/wm/background.vala [hunk 5]
```diff
@@ -140,6 +140,71 @@ public class BudgieBackground : Meta.BackgroundGroup
...
+        catch (Error e)
+        {
+            warning("Could not contact accounts service to look up '%s': %s", Environment.get_user_name(), e.message);
+            return;
+        }
+
+        string object_path = variant.get_child_value(0).get_string();
+
+        try
+        {
```
Comment (JoshStrobl) [inline src/wm/background.vala]: Trying to enforce a "new" (Josh) style everywhere, let's have curly braces on same line, single space before it :smile:

File: src/wm/background.vala [hunk 6]
```diff
@@ -140,6 +140,71 @@ public class BudgieBackground : Meta.BackgroundGroup
...
+                "FindUserByName",
+                new Variant("(s)", Environment.get_user_name()),
+                new VariantType("(o)"),
+                DBusCallFlags.NONE,
+                -1,
+                null
+            );
+        }
+        catch (Error e)
+        {
```
Comment (JoshStrobl) [inline src/wm/background.vala]: Trying to enforce a "new" (Josh) style everywhere, let's have curly braces on same line, single space before it :smile:

File: src/wm/background.vala [hunk 7]
```diff
@@ -140,6 +140,71 @@ public class BudgieBackground : Meta.BackgroundGroup
...
+            bus = Bus.get_sync(BusType.SYSTEM);
+        }
+        catch (IOError e)
+        {
+            warning("Failed to get system bus: %s", e.message);
+            return;
+        }
+
+        try
+        {
```
Comment (JoshStrobl) [inline src/wm/background.vala]: Trying to enforce a "new" (Josh) style everywhere, let's have curly braces on same line, single space before it :smile:

File: src/wm/background.vala [hunk 8]
```diff
@@ -140,6 +140,71 @@ public class BudgieBackground : Meta.BackgroundGroup
...
+     * to update the greeter background if the display
+     * manager supports the dbus call.
+     */
+    void set_accountsservice_user_bg(string background)
+    {
+        DBusConnection bus;
+        Variant variant;
+
+        try
+        {
```
Comment (JoshStrobl) [inline src/wm/background.vala]: Trying to enforce a "new" (Josh) style everywhere, let's have curly braces on same line, single space before it :smile:

File: src/wm/background.vala [hunk 9]
```diff
@@ -140,6 +140,71 @@ public class BudgieBackground : Meta.BackgroundGroup
         }
     }
 
+    /**
+     * call accountsservice dbus with the background file name
+     * to update the greeter background if the display
+     * manager supports the dbus call.
+     */
+    void set_accountsservice_user_bg(string background)
+    {
```
Comment (JoshStrobl) [inline src/wm/background.vala]: Trying to enforce a "new" (Josh) style everywhere, let's have curly braces on same line, single space before it :smile:

File: src/wm/background.vala [hunk 10]
```diff
@@ -140,6 +140,71 @@ public class BudgieBackground : Meta.BackgroundGroup
...
+                "org.freedesktop.Accounts",
+                "FindUserByName",
+                new Variant("(s)", Environment.get_user_name()),
+                new VariantType("(o)"),
+                DBusCallFlags.NONE,
+                -1,
+                null
+            );
+        }
+        catch (Error e)
```
Comment (JoshStrobl) [inline src/wm/background.vala]: Have catch on same line as curly brace, so `} catch (Error e) {`

File: src/wm/background.vala [hunk 11]
```diff
@@ -140,6 +140,71 @@ public class BudgieBackground : Meta.BackgroundGroup
...
+                            "org.freedesktop.DisplayManager.AccountsService",
+                            "BackgroundFile",
+                            new Variant.string( background )),
+                new VariantType("()"),
+                DBusCallFlags.NONE,
+                -1,
+                null
+            );
+        }
+        catch (Error e)
```
Comment (JoshStrobl) [inline src/wm/background.vala]: Have catch on same line as curly brace, so `} catch (Error e) {`

File: src/wm/background.vala [hunk 12]
```diff
@@ -140,6 +140,71 @@ public class BudgieBackground : Meta.BackgroundGroup
...
+    void set_accountsservice_user_bg(string background)
+    {
+        DBusConnection bus;
+        Variant variant;
+
+        try
+        {
+            bus = Bus.get_sync(BusType.SYSTEM);
+        }
+        catch (IOError e)
```
Comment (JoshStrobl) [inline src/wm/background.vala]: Have catch on same line as curly brace, so `} catch (Error e) {`

**Comment by JoshStrobl** [comment]
Could still probably do with more consolidation (IMO some needless early line breaking) but that's something I can sort out post merge :+1: Thanks for the pull request and new fanciness!

**Pull request merged**